textile/features.textile

```diff
@@ -1185,8 +1187,9 @@ h3(#activation-state-machine). Activation State Machine
 **** @(RSH3d1b)@ Transitions to @WaitingForNewPushDeviceDetails@.
 *** @(RSH3d2)@ On event @CalledDeactivate@:
 **** @(RSH3d2a)@ If a custom @deregisterCallback@ was provided to @Push#deactivate@, pass it the local @DeviceDetails@ 's id.
-**** @(RSH3d2b)@ Otherwise, make an asynchronous DELETE HTTP request to "/push/deviceRegistrations":/rest-api/#delete-device-registration using the local @DeviceDetails@ 's ID. This operation requires "push device authentication":#push-device-authentication.
+**** @(RSH3d2b)@ Otherwise, make an asynchronous DELETE HTTP request to "/push/deviceRegistrations":/rest-api/#delete-device-registration using the local @DeviceDetails@ 's ID. This operation requires "push device authentication":#push-device-authentication without other token or key authentication.
 **** @(RSH3d2c)@ Either way, when the registration is done, a @Deregistered@ or @DeregistrationFailed@ event should be fired.
+**** @(RSH3d2c1)@ @Deregistered@ event should be fired if the HTTP request returns a status code in the 2xx range, 401 (unauthorized), or code 40005 (invalid credentials). Otherwise @DeregistrationFailed@ event should be fired.
```

Comment:
this feels wrong to me, haven't fully thought about the changes but doesn't this mean that if a client temporarily has invalid auth credentials then it would appear as deregistered even if the device registration still exists?

Reply:
Deregistration call should relying solely on device identity token authentication (without any key or temporary token auth), as I highlighted in `RSH3d2b`. Device identity token is permanent and should never change after push registration. If it becomes invalid, it's impossible to update. We can narrow down to `40100` error code, we had race-condition problem before and created a lot of states where `deviceIdentityToken` is not align with `deviceId` so every time we are trying to deregister we get `"Token deviceId does not match"`, and the only solution right now is to reinstall the app. 

But I have feeling that it is much better to handle any given problem that can happen to `deviceIdentityToken`. So I added  `Invalid credentials` that you can get if your token is not well-formed or `Token unrecognised` that you can get if you provide key / token auth from another Ably app. This way we can be sure that even if there are some problem in SDK which resulted in broken state, we can deregister. Otherwise end-user have to reinstall the app.

Reply:
@owenpearson - Evegenii is off for this week. Are we happy to resolve this, or is there anything I can do to help get this in?

Reply:
@umair-ably, I'm still not sure I'm convinced, sorry. My question was more about what would happen if the client had invalid API auth credentials (ably key/token) rather than a problem with its local `deviceIdentityToken`. I haven't tested it, but from what I can understand, this change would mean that a device without valid API credentials would appear as if it has successfully deregistered even though its attempt to remove its own device registration is rejected by the API with a 401, which would be wrong since the registration would still exist in that case.

Reply:
Ahh I see what you mean. If my understanding is correct then something like this would play out... 

1. User successfully registers with valid auth 
2. User hits invalid auth error for whatever reason (ably key or token)
4. Server returns 401. The server does no processing itself since invalid auth was used.  
5. Device clears local registration state, allowing user to attempt registration again.
6. Server still assumes registered state

Seems like the risk here is that the server and client will be out of sync e.g. server assumes registered, client assumes deregistered. In this case, I think the changes to `RSH3d2b` in this commit would address that with something like this:

1. User successfully registers with valid auth 
2. User calls deregister with only device identity token (not the usual ably key or token)
4. Server returns 2xx if there is a matching and now removed device identify token, or a 40x (perhaps a 404 or 204 imo) if it could not find the token to begin with.
5. On both success or failure from step 4, the device clears local registration state, allowing the user to attempt registering again. (this is the critical step to fix the race conditions described by Evgenii above)
6. Server and client are in sync 

Have I missed something here?

Reply:
Yep, your understanding of the problem sounds correct 🙂 as for the solution, I'm not sure if you can send a `DELETE /push/deviceRegistrations` with just the deviceIdentityToken for authentication. I haven't tested it but I'd be quite surprised if that was the case. Moreover, if a client somehow ended up with an invalid deviceIdentityToken then you would likely end up with the same issue of client/server getting out of sync, although i guess this is quite an edge case since that would only really be caused by an sdk bug or a user sabotaging their own deviceIdentityToken.

I do think deleting local state upon a 404 makes sense, since in that case the server is actually saying that the device registration doesn't exist. Maybe if we're currently sending 401s on the basis that the deviceIdentityToken is no longer valid then we should change the API to respond with a 404 in that case?

Reply:
Evegenii put in a PR for ably-java before going on leave which also implies `RSH3d2b` works by calling delete solely with the deviceIdentityToken [here](https://github.com/ably/ably-java/pull/994). 

The deviceIdentityToken should never change once a user registers for push. Unless, as you said, a real edge case occurs in which we should delete it locally or else the user is stuck with it. 

> Maybe if we're currently sending 401s on the basis that the deviceIdentityToken is no longer valid then we should change the API to respond with a 404 in that case?

I don't think we are sending 401s right now on the basis they are no longer valid... it is purely down to invalid auth. The changes to `RSH3d2b` would allow us to do this though. 

I agree with yourself and Paddy that the wider design for this should mean `NOT ACTIVE` also means not persisting device details. There have been a few calls to simplify this state machine 😅 

I guess the context for why this solution was the one proposed was that it is currently the cause of several thousand broken states for end clients a day, which in turn means users are not receiving push notifications when they should be. I think this solution is adequate if my understanding of `RSH3d2b` is correct. Without `RSH3d2b`, there is a risk users think they've deregistered but they actually haven't (causing the inverse of current issues), however subsequent calls with valid auth would fix this (unlike now where this would still not resolve our issues)

Reply:
Just discussed this with @ttypic on a call, I had missed the fact that this request would only be sent with a deviceIdentityToken (ie excluding api credentials) so I'll resolve this thread now.